### ISSUE TITLE
error printing resources

### ISSUE THREAD

**Issue by erendrake** [KSP-KOS/KOS#813]
PRINT SHIP:RESOURCES AT (0, 0).

causes 

[Log]: System.ArgumentException: Destination array was not long enough. Check destIndex and length, and the array's lower bounds
  at System.Array.Copy (System.Array sourceArray, Int32 sourceIndex, System.Array destinationArray, Int32 destinationIndex, Int32 length) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Screen.ScreenBufferLine.ArrayCopyFrom (System.Char[] source, Int32 sourceStart, Int32 destinationStart, Int32 length) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Screen.ScreenBuffer.PrintLine (System.String textToPrint) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Screen.ScreenBuffer.Print (System.String textToPrint, Boolean addNewLine) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Screen.ScreenBuffer.PrintAt (System.String textToPrint, Int32 row, Int32 column) [0x00000] in <filename unknown>:0 
  at kOS.Screen.Interpreter.PrintAt (System.String textToPrint, Int32 row, Int32 column) [0x00000] in <filename unknown>:0 
  at kOS.Function.FunctionPrintAt.Execute (kOS.SharedObjects shared) [0x00000] in <filename unknown>:0 
  at kOS.Function.FunctionManager.CallFunction (System.String functionName) [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.CallBuiltinFunction (System.String functionName) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Compilation.OpcodeCall.Execute (ICpu cpu) [0x00000] in <filename unknown>:0 
  at kOS.Execution.CPU.ExecuteInstruction (kOS.Execution.ProgramContext context) [0x00000] in <filename unknown>:0

**Comment by Dunbaratu** [comment]
huh?
What is the ToString() of SHIP:RESOURCES ?
is it an empty string?

I've seen problems before trying to ```print "".```, where ```print " ".``` works just fine.

**Comment by erendrake** [comment]
I think i have spotted it.

**Comment by erendrake** [comment]
nm i dont.

    PRINT SHIP:RESOURCES. 

works fine, just not
    PRINT SHIP:RESOURCES at (0, 0).

**Comment by Dunbaratu** [comment]
not near my computer at the time.

Is it just ship resources or does any string at position zero have the problem?

**Comment by erendrake** [comment]
LIST ENGINES IN elist.
    PRINT elist AT (0,0).

gives the same error.

**Comment by erendrake** [comment]
the following script errors every time 

    SET newList to LIST().

    CLEARSCREEN.
    PRINT newList AT (0,0).

    WAIT 1.

    CLEARSCREEN.
    SET xLong TO "Test                                                                                            asdfasdf                                                                     Hi!".

    PRINT xLong AT (0,0).

    WAIT 1.

    CLEARSCREEN.
    newList:ADD(xLong).
    PRINT newList AT (0,0).

    WAIT 1.

**Comment by erendrake** [comment]
i have a hunch that this might be because printat cannot start a new line??

https://github.com/KSP-KOS/KOS/blob/6b7d217b11dfed690da105d369083d1a1da1c220/src/kOS.Safe/Screen/ScreenBuffer.cs#L130

**Comment by abenkovskii** [comment]
@erendrake I can confirm the cause of the bug is that `print at` can't print long strings with LF characters. By opcode editing I managed to get a long multi line string. And than I got the same error when trying to `print at` it.

**Comment by Dunbaratu** [comment]
Only breaks if there is a newline embedded in the string being printed, and the newline occurs while the cursor is past the right edge of the terminal.

This reliably triggers it:
```
// To trigger the bug, these strings in the LIST need to be wider than the terminal screen:
set foo to LIST(
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
  "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
).

// This works fine:
print foo.

// Whereas this triggers the bug:
print foo at (0,10).

```
It requires ALL of the following to be true:
- The printout is wider than the terminal.
- The string being printed contains newlines.
- The print is using the "at (x,y)" system, rather than just being a normal print.

**Comment by Dunbaratu** [comment]
I've fixed this but in the process I discovered that there's a second bug:  Using PRINT AT causes the printout to utterly ignore newlines and instead join together the lines, like so:

```
set foo to LIST("aaa","bbb").
```
Printing it normally does this:
```
print foo.
List of 2 items:
 [ 0]=aaa
 [ 1]=bbb
```
While printing it with "at" joins it all on one line:
```
print foo at (0.3).

List of 2 items: [ 0]=aaa [ 1]=bbb
```
I don't think this is necessarily a problem, but it should be noted and maybe refactored later.  The problem is we'd have to define what the meaning is of the second such line.  i.e. if you do:

print foo at (10,2).

Should that look like this:
```

          List of 2 items:
 [ 0]=aaa
 [ 1]=bbb
```
or like this:
```

          List of 2 items:
           [ 0]=aaa
           [ 1]=bbb
```
(does the start position represent a relative indent that all subsequent lines share or not?)

But that's for the future maybe in another issue.  I've fixed the crash and will post a PR about just that.
